x/stakeibc/keeper/msg_server.go

```diff
@@ -1103,3 +1106,66 @@ func (k msgServer) ResumeHostZone(goCtx context.Context, msg *types.MsgResumeHos
 
 	return &types.MsgResumeHostZoneResponse{}, nil
 }
+
+// Submits an ICA tx to either grant or revoke authz permisssions to an address
+// to execute trades on behalf of the trade ICA
+func (k msgServer) ToggleTradeController(
+	goCtx context.Context,
+	msg *types.MsgToggleTradeController,
+) (*types.MsgToggleTradeControllerResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Fetch the trade ICA which will be the granter
+	tradeRoute, found := k.GetTradeRouteFromTradeChainId(ctx, msg.ChainId)
+	if !found {
+		return nil, types.ErrTradeRouteNotFound.Wrapf("trade route not found for chain ID %s", msg.ChainId)
+	}
+
+	swapMsgTypeUrl := "/" + proto.MessageName(&types.SwapAmountInRoute{})
+
+	// Build the authz message, depending on whether permissions are granted or revoked
+	// The granter will always be the trade ICA and the grantee will be the address
+	// specified in the message
+	var authzMsg []proto.Message
+	switch msg.PermissionChange {
+	case types.AuthzPermissionChange_GRANT:
+		authzMsg = []proto.Message{&authz.MsgGrant{
+			Granter: tradeRoute.TradeAccount.Address,
+			Grantee: msg.Address,
+			Grant:   authz.Grant{
+				// TODO [rebate]
+			},
+		}}
+	case types.AuthzPermissionChange_REVOKE:
+		authzMsg = []proto.Message{&authz.MsgRevoke{
+			Granter:    tradeRoute.TradeAccount.Address,
+			Grantee:    msg.Address,
+			MsgTypeUrl: swapMsgTypeUrl,
+		}}
+	default:
+		return nil, errors.New("invalid permission change")
+	}
+
+	// Build the ICA channel owner from the trade route
+	tradeRouteAccountOwner := types.FormatTradeRouteICAOwnerFromRouteId(
+		msg.ChainId,
+		tradeRoute.GetRouteId(),
+		types.ICAAccountType_CONVERTER_TRADE,
+	)
+
+	// Submit the ICA tx from the trade ICA account
+	// Timeout the ICA at 1 hour
+	timeoutTimestamp := uint64(ctx.BlockTime().Add(time.Hour).UnixNano())
+	err := k.SubmitICATxWithoutCallback(
+		ctx,
+		tradeRoute.TradeAccount.ConnectionId,
+		tradeRouteAccountOwner,
+		authzMsg,
+		timeoutTimestamp,
+	)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
When you run this, the memory of which permissions which addresses currently have is on the other chain where the ICA exists, but you could store an array of which addresses currrently are granted or revoked.  Then instead of the PR which just removes the hook for the Trade to happen on chain, it could check if there are any operators granted permissions and it would only do the epochly on chain trades if there are no current operators with permissions.  Sort of a fall back fail safe and allows us to turn on or off operator mode very easily.

Reply:
that's an interesting idea! that would be a nice fallback, but I'm not sure having a fallback is valuable enough to justify making this process stateful. Curious to hear other people's thoughts as well

Reply:
I agree, I think the additional stateful complexity isn't worth the auto-fall-back on-chain! Especially given the tight timeline, I'd bias to just shipping the simple thing and we can add the additional complexity later